### ISSUE TITLE
Is "Case statements should not only contain `true` and `false`." a good default?

### ISSUE THREAD

**Issue by fishcakez** [rrrene/credo#65]
Suggesting an `if..else` is less strict, which could potentially hide bugs because `if` tests for truthy rather than `true`.

``` elixir
case Map.fetch!(map, key) do
  true -> :good
  false -> :bad
end
```

Situations where one might use a `case` like this are when a map or keyword have been passed from another application.

**Comment by rrrene** [comment]
Mmmmh, that's a compelling argument. :confused: 

My counter argument would be that your example could also be written as:

    if Map.fetch!(map, key) == true do

Where I would explicitly check for `true` (a use case I encounter frequently myself) and handle all non-true cases in an `else` block.

As everything in Credo this is pretty much up for debate. The main reason for this check was easier readability for simple control flows, which don't require pattern matching. I think `if/else` is more easily understood by beginners (and thus "people != you" maintaining your code in the future).

**Comment by fishcakez** [comment]
Yeah I understand credo defaults are up for debate and it is a matter of opinion. I tried credo on a project and got about 130 warnings/suggestions and I disagreed with all but 1 suggestion :/. I think about 12 uniques. This was the only one that I thought was counterproductive. The others I could sort of understand someone wanting but weren't for me.

> if Map.fetch!(map, key) == true do

That that isn't equivalent to the case above, a truthy other than `true` would be treated as false. A truthy test seems simple but actually has a lot more going on than the case. So a beginner can think they understand but not actually.

I am not a big fan of using `if/else` unless testing for a truthy or testing on a function that always returns a boolean, e.g. `is_X` or `x?`. I find the lack of explicitness awkward and a potential spot for hiding bugs. This shows our opinions on this may be so far from each other we may have to agree to disagree :smile:

**Comment by rrrene** [comment]
By "up for debate" I primarily meant that they are still changing/evolving. I see what you are getting at with your argument and the fact that `if` is less explicit than `case` is something I did not consider. 

> got about 130 warnings/suggestions and I disagreed with all but 1 suggestion

Wow. While I recognize lots of stuff is bound to opinion and some checks might be controversial, I did not see that coming. What was the single suggestion you agreed with?

While a tool like Credo will always be controversial, I still want to ensure we have sensible defaults that most people can agree upon (or at least see value in while disagreeing). Since I see your concern regarding the counter-productiveness of the `CaseTrivialMatches` check, I would like to hear your thoughts on other (potentially misguided) checks as well :+1:

**Comment by rrrene** [comment]
Oh, I accidentally skipped over "This was the only one that I thought was counterproductive." :grin: 

Anyhow: If you have *any* other thoughts on this tool, I'd love to hear them!

**Comment by rrrene** [comment]
@fishcakez The check is now deprecated in `v0.4.0-beta1`. Thx again for discussing this with me! :+1:

**Comment by ijunaidfarooq** [comment]
```
  defp valid_schedule(key, json) do
    case is_map(json) do
      true -> :ok
      false -> invalid(key)
    end
  end
```
as per credo on this def its giving me the Refactorting Opportunites

```
 [F] → Case statements should not only contain `true` and `false`.
┃       web/validations/cloud_recording.ex:38 (EvercamMedia.Validation.CloudRecording.valid_schedule)
┃ [F] → Case statements should not only contain `true` and `false`.
┃       web/validations/cloud_recording.ex:38 (EvercamMedia.Validation.CloudRecording.valid_schedule)
┃ [F] → Case statements should not only contain `true` and `false`.
┃       web/validations/cloud_recording.ex:38 (EvercamMedia.Validation.CloudRecording.valid_schedule)

```
as why three times?

**Comment by rrrene** [comment]
That's not good. I will have to check!

**Comment by ijunaidfarooq** [comment]
![selection_031](https://cloud.githubusercontent.com/assets/13561999/15419919/e6af8010-1e82-11e6-9641-dfb13ebc5f5d.png)

for your more convenience! this is what happened.
what I did was `mix credo web/validations/cloud_recording.ex`
and its said `Checking 3 source files ...`

**Comment by rrrene** [comment]
o_O damn. Seems to be a deeper issue not strictly related to CaseTrivialMatches.

**Comment by ijunaidfarooq** [comment]
Also what's with the duplicate code found at `defmodule EvercamMedia.Validation.CloudRecording do` O_o ? 

FYI. am very beginner to this!

**Comment by rrrene** [comment]
I'm sorry, but it seems I can not reproduce this:
![image](https://cloud.githubusercontent.com/assets/311914/15454959/961dda0a-2046-11e6-9758-5f98d1a267a0.png)
You are on OS X, right?

**Comment by ijunaidfarooq** [comment]
Nope.. I am on ubuntu 14.04..
